Framework/Algorithms/src/ConvertUnitsUsingDetectorTable.cpp

```diff
@@ -276,29 +120,52 @@ void ConvertUnitsUsingDetectorTable::convertViaTOF(
         "DetectorParameter TableWorkspace is not defined correctly.");
   }
 
-  // Now let's read them into some vectors.
-  auto l1Column = paramWS->getColVector<double>("l1");
-  auto l2Column = paramWS->getColVector<double>("l2");
-  auto twoThetaColumn = paramWS->getColVector<double>("twotheta");
-  auto efixedColumn = paramWS->getColVector<double>("efixed");
-  auto emodeColumn = paramWS->getColVector<int>("emode");
-  auto spectraColumn = paramWS->getColVector<int>("spectra");
-
-  EventWorkspace_sptr eventWS =
-      boost::dynamic_pointer_cast<EventWorkspace>(outputWS);
-  assert(static_cast<bool>(eventWS) == m_inputEvents); // Sanity check
+  // Now let's take a reference to the vectors.
+  const auto &l1Column = paramWS->getColVector<double>("l1");
+  const auto &l2Column = paramWS->getColVector<double>("l2");
+  const auto &twoThetaColumn = paramWS->getColVector<double>("twotheta");
+  const auto &efixedColumn = paramWS->getColVector<double>("efixed");
+  const auto &emodeColumn = paramWS->getColVector<int>("emode");
+  auto &spectraColumn = paramWS->getColVector<int>("spectra");
 
   Progress prog(this, 0.2, 1.0, m_numberOfSpectra);
   int64_t numberOfSpectra_i =
       static_cast<int64_t>(m_numberOfSpectra); // cast to make openmp happy
 
   // Get the unit object for each workspace
-  Kernel::Unit_const_sptr outputUnit = outputWS->getAxis(0)->unit();
-
+  Kernel::Unit_const_sptr outputUnit = m_outputUnit;
   std::vector<double> emptyVec;
   int failedDetectorCount = 0;
 
-  // ConstColumnVector<int> spectraNumber = paramWS->getVector("spectra");
+  // Perform Sanity Validation before creating workspace
```

Comment:
The previous code had some checks but they are down in the heart of the ToTOF and FromTOF of the unit classes themselves.  As such they were the cause of the problem as they threw exceptions after the output workspaces had already been modified.  Here, as in ConvertUnits I perform a trial unit conversion of a single spectra (from the input workspace) before I create the output workspace, so any errors from the unit classes can fire safely.

A better design would be to separate this validation into a more explicit method, but it would still need pretty much the same inputs.  Given the Core nature of ConvertUnits, I went for the smaller change.